engines/e_pkcs11_eng.c

```diff
@@ -0,0 +1,598 @@
+
+    if (ctx->pin == NULL) {
+        if (!pkcs11_get_console_pin(&pin))
+            goto err;
+        ctx->pin = (CK_BYTE *) pin;
+        if (ctx->pin == NULL) {
+            PKCS11_trace("PIN is invalid\n");
+            goto err;
+        }
+    }
+    return 1;
+
+ err:
+    return 0;
+}
+
+static EVP_PKEY *pkcs11_engine_load_private_key(ENGINE * e, const char *path,
+                                                UI_METHOD * ui_method,
+                                                void *callback_data)
+{
+    CK_RV rv;
+    PKCS11_CTX *ctx;
+
+    ctx = ENGINE_get_ex_data(e, pkcs11_idx);
+    if (!pkcs11_parse(ctx, path))
+        goto err;
+
+    rv = pkcs11_initialize(ctx->module_path);
+
+    if (rv != CKR_OK)
+        goto err;
+    if (!pkcs11_get_slot(ctx))
+        goto err;
+    if (!pkcs11_start_session(ctx))
+        goto err;
+    if (!pkcs11_login(ctx, CKU_USER))
+        goto err;
+    if (!pkcs11_find_private_key(ctx))
+        goto err;
+
+    return pkcs11_load_pkey(ctx);
+
+ err:
+    PKCS11_trace("pkcs11_engine_load_private_key failed\n");
+    return 0;
+}
+
+static OSSL_STORE_LOADER_CTX* pkcs11_store_open(
+    const OSSL_STORE_LOADER *loader, const char *uri,
+    const UI_METHOD *ui_method, void *ui_data)
+{
+    ENGINE *e;
+    PKCS11_CTX *pkcs11_ctx;
+    OSSL_STORE_LOADER_CTX *store_ctx = NULL;
+    char *objecttype = NULL, *object = NULL;
+
+    store_ctx = OSSL_STORE_LOADER_CTX_new();
+
+    e = (ENGINE *) OSSL_STORE_LOADER_get0_engine(loader);
+    if (e == NULL)
+        return NULL;
+
+    uri += 7;
+
+    pkcs11_parse_uri(uri,"objecttype=", &objecttype);
+    pkcs11_parse_uri(uri,"object=", &object);
+
+    if (object != NULL && objecttype != NULL
+        && strncmp(objecttype, "cert", 4) == 0) {
+        pkcs11_ctx = ENGINE_get_ex_data(e, pkcs11_idx);
+        if (pkcs11_ctx == NULL)
+            return NULL;
+        if (!pkcs11_get_cert(store_ctx, pkcs11_ctx, object))
+            return NULL;
+    }
+
+    if (strncmp(uri, "list=all", 8) == 0) {
+        pkcs11_ctx = ENGINE_get_ex_data(e, pkcs11_idx);
+        if (pkcs11_ctx == NULL)
+            return NULL;
+        if (!pkcs11_get_ids(store_ctx, pkcs11_ctx))
+            return NULL;
+    }
+
+    return store_ctx;
+}
+
+static OSSL_STORE_INFO* pkcs11_store_load(OSSL_STORE_LOADER_CTX *ctx,
+                                          const UI_METHOD *ui_method,
+                                          void *ui_data)
+{
+    OSSL_STORE_INFO *ret = NULL;
+    char *name;
+    char *description;
+
+    if (ctx->cert != NULL) {
+        ctx->eof = 1;
+        return pkcs11_store_load_cert(ctx, ui_method, ui_data);
+    }
+
+    if (ctx->ids[store_idx].name[0] == '\0' || store_idx == MAX - 1) {
+        ctx->eof = 1;
+        return NULL;
+    }
+
+    name = OPENSSL_strdup(ctx->ids[store_idx].name);
+    description = OPENSSL_strdup(ctx->ids[store_idx].desc);
+    ret = OSSL_STORE_INFO_new_NAME(name);
```

Comment:
Wellllll....  not quite.

This section of the [`OSSL_STORE_INFO`](https://github.com/openssl/openssl/blob/master/doc/man3/OSSL_STORE_INFO.pod) manual is very relevant:

> **OSSL_STORE_INFO_NAME**
> 
> A name is exactly that, a name. It's like a name in a directory, _but formatted as a complete URI_. For example, the path in URI file:/foo/bar/ could include a file named cookie.pem, and in that case, the returned OSSL_STORE_INFO_NAME object would have the URI file:/foo/bar/cookie.pem, which can be used by the application to get the objects in that file. This can be applied to all schemes that can somehow support a listing of object URIs.
> 
> For file: URIs that are used without the explicit scheme, the returned name will be the path of each object, so if /foo/bar was given and that path has the file cookie.pem, the name /foo/bar/cookie.pem will be returned.
> 
> The returned URI is considered canonical and must be unique and permanent for the storage where the object (or collection of objects) resides. Each loader is responsible for ensuring that it only returns canonical URIs. However, it's possible that certain schemes allow an object (or collection thereof) to be reached with alternative URIs; just because one URI is canonical doesn't mean that other variants can't be used.
> 
> At the discretion of the loader that was used to get these names, an extra description may be attached as well.

Basically, the given name *must* be possible to use as an individual URI to get the object in question with a separate OSSL_STORE loading session.